Library/Homebrew/api/formula.rb

```diff
@@ -101,6 +102,71 @@ def write_names_and_aliases(regenerate: false)
             end
           end
         end
+
+        REQUIRED_FIELDS = %w[
```

Comment:
I know this blows up the scope a bit but: I'd instead be tempted to actually have a new `Formula#to_hash` (and Cask equivalent) or pass a versioned argument there and generate an entirely new hash format with e.g. less nesting/repetition that better meets our needs.

May be a good time to actually have some sort of stricter versioning or e.g. specifying a minimum Homebrew version in the JSON itself here, too.

Reply:
That would be ideal but I'm a bit worried about backwards compatibility. I'm not sure how many people depend on the output of `brew info --json=v2` though. I guess that could just be json v3 and we could just keep the old format at json v2.

Also, changing the format entirely would add a lot of complexity to the load from API methods which rely on the current format. This is especially important for formula that get removed since we'll have to load them from the original JSON definition.

Reply:
I guess we could have multiple loaders and just have a key like `json_version` that specifies which one we should try and load it with.

Reply:
Here's a sample of the current slimmed-down JSON representation. Let me know if there's anything that looks like it can be removed and/or simplified.

https://gist.github.com/apainintheneck/882bb75d29394f7bdd5f8b5efe482038

Reply:
`"versions" -> "head"` can probably be trimmed.

It's quite an interesting one that. In theory we might support:

```ruby
head do
  url "..."
  version "whatever"
end
```

In practice: does it actually work? Not that it's a concern for Homebrew/core, but maybe something we could make error on a Ruby DSL level.

Reply:
Looks like it does! https://github.com/dart-lang/homebrew-dart/blob/d433a4de95e9f361203c249847fc384fbf47b840/Formula/dart.rb#L9

Not a concern for Homebrew/core in any case - we can have a cop to disallow it for Homebrew/core tap specifically (unlesss we already do)

Reply:
That's a good example of something we can likely remove because it doesn't fit the core repo use case.

Looking at it a little more, we could potentially remove a few more by omitting values if they correspond to a default.

For example, the following fields all seem to default to 0 so we could just omit them in that case and add 0 manually when loading things.

```rb
json["bottle"]["stable"]["rebuild"] == 0
json["revision"] == 0
json["version_scheme"] == 0
```

Also at a glance the versions hash always seems to include `"bottle": true` for core formulas anyway. Not sure if we need to include that.

The tap information is consistent per file and should only need to be stored once though it complicates things a bit when deserializing things.

```rb
json["tap"]
json["tap_git_head"]
json["bottle"]["stable"]["root_url"]
```

Reply:
> The tap information is consistent per file and should only need to be stored once

Yes what we should have done is have a metadata section at the start but we picked a flat array so can't do that.

Reply:
We could still do something like this.

```rb
{
  "payload": {
    "metadata": {...},
    "version": 2,
    "formulae": [...]
  }
}
```

It'd be easy enough to try later on but doesn't necessarily have to make it into this PR.

Reply:
Yeah in my mind a theoretical tap agnostic format would have been:

```
{
  "tap_info": {
    ...,
    "migrations": [...]
  },
  "formulae": [...],
  "casks": [...]
}
```

The tap_migrations JSON being a separate download is a hack indicative of this problem. All of this is backwards incompatible changes however. We're already going to break backwards compatibility here in terms of old brew versions but not break backwards compatibility in terms of old JSON data (though that can do so if we do what we did when moving to JWS).

Reply:
I'm not sure I understand your comment about https://github.com/Homebrew/brew/commit/984dcf8c11a571871f278c9a39c099b85f75066f. Those changes should be backwards compatible.

I like your idea about the tap agnostic format. I think it should be possible to implement that in a backwards compatible way though. Basically, we could add additional logic to build the `Homebrew::API::Formula.all_formulae` hash so that it stays the same as before minus the missing hash keys despite the changes to the `formula.jws.json` file. Whether that's worth the effort is another question entirely. This would mean that the JSON files saved during the installation should hold the same amount of information as before.

What was the goal of the dependency redo code you mention here?

Reply:
So we'd have two different branches of parsing logic for the old and new `formula.jws.json` files but the same resulting `Homebrew::API::Formula.all_formulae` hash if that makes sense.

Reply:
> I guess that could just be json v3 and we could just keep the old format at json v2.

Yes, sorry, this was my thinking 👍🏻. Might even be something like `vHomebrewCore` or something as well as a separate v3 which could have more information in a nicer structure.

> That's a good example of something we can likely remove because it doesn't fit the core repo use case.

Yeh, this is my thinking: we remove anything that we won't use/need in core.

> For example, the following fields all seem to default to 0 so we could just omit them in that case and add 0 manually when loading things.
> ...
> Also at a glance the versions hash always seems to include `"bottle": true` for core formulas anyway. Not sure if we need to include that.

These seem like good things to omit.

> but this would be a compatibility break in terms of old cached or mirrored JSON data (though we can do what we did when moving to JWS if we want to go this direction).

I think we should aim to just force a new API download when we release a new enough Homebrew version. I can see an argument for a couple of point releases where we have branching logic here but it's a small enough download that needs refetched/updated often enough I don't think it's worth keeping around for long at all beyond testing/gradual rollout.

> Whether that's worth the effort is another question entirely.

It is not 😁

Reply:
in the generated smaller json I see many `sha256` defined, specific always a list for many os. which of the other os I would need if I can download a pre-filtered jsons like `slim-formula-{os}.json`:

- `slim-formula-arm64_sonoma.json`
- `slim-formula-arm64_monterey.json`